File: frontend/src/lib/monitoring.ts
/**
 * Suivi des erreurs du navigateur avec Sentry (ADR 0013), inactif sans `NEXT_PUBLIC_SENTRY_DSN`.
 *
 * Seules les erreurs partent, et rien qui identifie l'utilisateur ou ouvre son compte : ni cookies, ni
 * adresse IP, ni le jeton des liens reçus par e-mail (`/reset-password?token=…`), que l'adresse de la page
 * porte et que Sentry joindrait sinon au rapport et à son fil d'événements.
 */
import type { Breadcrumb, ErrorEvent } from "@sentry/browser";

/** Adresse sans ses paramètres sensibles : `?token=abc` devient `?token=[retiré]`. */
export function withoutSecrets(url: string | undefined): string | undefined {
  if (!url) return url;
  return url.replace(/([?&](?:token|password)=)[^&#\s]*/gi, "$1[retiré]");
}

export function scrubEvent(event: ErrorEvent): ErrorEvent {
  if (event.request) {
    event.request.url = withoutSecrets(event.request.url);
    delete event.request.cookies;
    delete event.request.query_string;
    const userAgent = event.request.headers?.["User-Agent"];
    event.request.headers = userAgent ? { "User-Agent": userAgent } : {};
  }
  delete event.user;
  event.breadcrumbs = event.breadcrumbs?.map(scrubBreadcrumb);
  return event;
}

export function scrubBreadcrumb(breadcrumb: Breadcrumb): Breadcrumb {
  const data = breadcrumb.data;
  if (data) {
    for (const key of ["url", "from", "to"]) {
      if (typeof data[key] === "string") data[key] = withoutSecrets(data[key]);
    }
  }
  return breadcrumb;
}

export async function initMonitoring(dsn: string | undefined) {
  if (!dsn) return;
  // Chargé seulement s'il sert : sans DSN, le module ne quitte jamais le serveur
  const Sentry = await import("@sentry/browser");
  Sentry.init({
    dsn,
    environment: process.env.NODE_ENV === "production" ? "production" : "development",
    // Rien sur l'utilisateur, ni cookies, ni en-têtes, ni corps, ni paramètres d'adresse : scrubEvent le
    // vérifie encore avant l'envoi, au cas où une intégration en ajouterait
    dataCollection: { userInfo: false, cookies: false, httpHeaders: false, httpBodies: [], urlQueryParams: false },
    tracesSampleRate: 0,
    beforeSend: scrubEvent,
    beforeBreadcrumb: scrubBreadcrumb,
  });
}

File: backend/monitoring.py
# DANS backend/monitoring.py
"""
Suivi des erreurs avec Sentry ([ADR 0013](../docs/adr/0013-cible-hebergement-production.md)).

Inactif sans `SENTRY_DSN` : rien ne part en développement ni dans les tests. En production, seules les
erreurs sont envoyées (pas de mesure de performance : l'offre gratuite a un quota), et rien de ce qui
identifie un utilisateur ou ouvre son compte : ni cookies, ni en-têtes d'authentification, ni corps de
requête (mots de passe, jetons des liens reçus par e-mail), ni variables locales, ni adresse IP.
"""

import sentry_sdk
from flask import Flask
from sentry_sdk.integrations.flask import FlaskIntegration

# En-têtes gardés dans un rapport d'erreur : ceux qui aident à comprendre, aucun qui authentifie
KEPT_HEADERS = {"content-type", "content-length", "user-agent", "accept", "accept-language", "origin", "referer"}


def scrub(event: dict, _hint: dict) -> dict:
    """Retire d'un rapport ce qui pourrait identifier un utilisateur ou ouvrir son compte."""
    request = event.get("request")
    if request:
        request.pop("cookies", None)
        request.pop("data", None)
        request.pop("env", None)
        headers = request.get("headers") or {}
        request["headers"] = {k: v for k, v in headers.items() if k.lower() in KEPT_HEADERS}
    event.pop("user", None)
    return event


def init_sentry(app: Flask) -> bool:
    """Active Sentry si `SENTRY_DSN` est défini. Renvoie True s'il l'est."""
    dsn = app.config.get("SENTRY_DSN")
    if not dsn:
        return False
    sentry_sdk.init(
        dsn=dsn,
        environment=app.config.get("APP_ENV", "development"),
        release=app.config.get("RELEASE") or None,
        integrations=[FlaskIntegration()],
        send_default_pii=False,
        max_request_body_size="never",
        # Les variables locales d'une fonction qui plante peuvent contenir un mot de passe (le corps d'une
        # connexion validé dans `payload`, par exemple) : le filtre de Sentry ne les reconnaît qu'à leur nom
        include_local_variables=False,
        traces_sample_rate=0,
        before_send=scrub,
    )
    return True


def report_exception(error: BaseException) -> None:
    """Signale une exception gérée par l'API.

    Le gestionnaire d'erreurs de security.py intercepte toutes les exceptions pour renvoyer un 500 propre :
    Flask ne les voit plus passer, et l'intégration Sentry non plus. Sans Sentry actif, ne fait rien.
    """
    sentry_sdk.capture_exception(error)
